chainerx_cc/chainerx/device.cc

```diff
@@ -90,30 +106,42 @@ void GenericBatchNormForwardBackward::SetForwardResults(Array x, Array gamma, Ar
     gamma_ = std::make_shared<Array>(std::move(gamma));
     x_mean_ = std::make_shared<Array>(std::move(x_mean));
     x_inv_std_ = std::make_shared<Array>(std::move(x_inv_std));
+    beta_dtype_ = beta_dtype;
 }
 
 Array GenericBatchNormForwardBackward::Forward(const Array& x, const Array& gamma, const Array& beta) {
     CHAINERX_ASSERT(internal::GetArrayBody(x)->nodes().empty());
     CHAINERX_ASSERT(internal::GetArrayBody(gamma)->nodes().empty());
     CHAINERX_ASSERT(internal::GetArrayBody(beta)->nodes().empty());
 
-    Array x_mean = Mean(x, axis_, true);
-    Array x_var = Var(x, axis_, true);
+    CHAINERX_ASSERT(GetKind(x.dtype()) == DtypeKind::kFloat);
+    CHAINERX_ASSERT(GetKind(gamma.dtype()) == DtypeKind::kFloat);
+    CHAINERX_ASSERT(GetKind(beta.dtype()) == DtypeKind::kFloat);
+    CHAINERX_ASSERT(GetKind(running_mean().dtype()) == DtypeKind::kFloat);
+    CHAINERX_ASSERT(GetKind(running_var().dtype()) == DtypeKind::kFloat);
 
-    ApplyBatchNormResult result = ApplyBatchNorm(x, gamma, beta, x_mean, x_var, eps_, axis_);
+    // Compute the mean and variance of x with promoted dtype if the parameters have higher precisions.
+    Dtype dtype = ResultType(x, gamma, beta);
```

Comment:
Let's give a more descriptive name, such as `interm_dtype`, because one might think it is the final output dtype.